### PR TITLE
LSP infra: runtime-attached navigation query channel + lsp.delegate_to_runtime flag (BT-2239)

### PR DESCRIPTION
**Pull request by jamesc**
## Summary

Foundation for delegating LSP navigation queries to a running runtime's `SystemNavigation` facade — the static-first, live-augmented tooling model of ADR 0024. When a runtime is attached, the editor can answer find-references / go-to-implementation / hierarchy queries from the **live image**, which sees open-class extension methods (ADR 0066) and REPL-loaded classes the static AST walker cannot.

This is the foundation issue for the BT-2215 epic: it defines the long-lived wire contract and the reusable dispatch seam the per-method children (BT-2240..2244) build on. **No navigation behaviour changes by default** — the `lsp.delegate_to_runtime` flag is off, so all queries use the existing AST path.

Linear: https://linear.app/beamtalk/issue/BT-2239

## Key changes

- **Protocol** (`beamtalk-repl-protocol`): structured `nav-query` op (`implementorsOf` / `sendersOf` / `referencesTo`) returning machine-readable `{class, selector, line}` sites or `{class, meta}` implementors, with shared `NavSite` / `NavImplementor` response types.
- **Runtime** (`beamtalk_repl_ops_nav.erl`): routes the query to `SystemNavigation` in the calling eval worker (where these sealed-Object value-type methods run inline) and serialises via the shared `term_to_json` class formatter. The argument is whitelisted to selector / class-name characters so it cannot inject into the evaluated expression.
- **Core** (`language_service`): a pure, tested source-location translation helper (`method_relative_line_span` + `locate_nav_site`) mapping the runtime's method-relative `line` to an absolute `Location`, plus a `selector_at` accessor.
- **LSP**: a `tokio-tungstenite` `RuntimeClient` that discovers an attached runtime via `beamtalk-workspace` and speaks the authenticated REPL protocol (ADR 0020); the `delegateToRuntime` flag; and a reusable two-mode dispatch seam wired into find-references as the worked example.
- **Docs**: `repl-protocol.md` (new op) + `surface-parity.md`.

Reuses `beamtalk-repl-protocol` and `beamtalk-workspace` so `beamtalk-lsp` never depends on `beamtalk-cli` (architecture constraint).

## Scope note

find-references is the worked example (selector senders only); full references parity (class references, edge cases, snapshot tests) and the other navigation methods land in the BT-2215 children, which now have a working transport + seam to build on.

## Test plan

- [x] `beamtalk-repl-protocol` unit tests (nav-query request + NavSite/NavImplementor parsing)
- [x] `beamtalk-core` unit tests (method-relative line span + translation, incl. edge cases)
- [x] `beamtalk-lsp` unit tests (flag default-off, flag reader, site→location translation, unknown-class skip)
- [x] `beamtalk_repl_ops_nav_tests` EUnit (13 tests: validation, injection guard, serialisation shaping)
- [x] clippy, rustfmt, erlfmt, dialyzer, dialyzer-specs all clean
- [ ] End-to-end LSP↔runtime↔SystemNavigation exercised by BT-2240 (needs a live runtime harness)

https://claude.ai/code/session_019miSbTZVfEK86ew7JxFw7b

---
_Generated by [Claude Code](https://claude.ai/code/session_019miSbTZVfEK86ew7JxFw7b)_

## Summary by CodeRabbit

* **New Features**
  * Runtime-delegated navigation: LSP can query an attached runtime for references/implementors and return accurate editor locations.
  * WebSocket runtime client with auth, per-query timeouts, and automatic reconnect on failure.
  * Improved on-cursor selector detection and mapping from runtime-relative lines to source locations.

* **Documentation**
  * Added navigation operations and updated REPL/protocol docs describing the new nav-query.

* **Tests**
  * New unit and integration tests for runtime nav requests, protocol shapes, encoding, and validation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/jamesc/beamtalk/pull/2296?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: crates/beamtalk-lsp/src/runtime.rs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            .await
+            .map_err(|_| "runtime query timed out".to_string())?
+    }
+}
+
+/// Perform the ADR 0020 auth handshake: read `auth-required`, send the cookie,
+/// read `auth_ok`, read `session-started`.
+async fn perform_auth_handshake(ws: &mut WsStream, cookie: &str) -> Result<(), String> {
+    let auth_required = read_text_message(ws).await?;
+    let parsed: serde_json::Value =
```
[the remark this answers is not in the document]
Comment (jamesc) [reply crates/beamtalk-lsp/src/runtime.rs:131]: Fixed in 2af9779 — the auth handshake is now wrapped in `tokio::time::timeout(CONNECT_TIMEOUT, perform_auth_handshake(...))`, so a runtime that stalls after connect returns an error and the caller falls back to the AST path instead of hanging.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: runtime/apps/beamtalk_workspace/src/beamtalk_repl_ops_nav.erl [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    Impls = [encode_implementor(C) || C <- Term],
+    Base = beamtalk_repl_protocol:base_response(Msg),
+    Result = Base#{<<"implementors">> => Impls, <<"status">> => [<<"done">>]},
+    iolist_to_binary(json:encode(Result));
+encode_result(_Shape, _Term, Msg) ->
+    %% Defensive: SystemNavigation always returns a List; a non-list result
+    %% means the query shape changed. Surface an empty result rather than crash.
+    ?LOG_WARNING("nav-query: unexpected non-list result", #{domain => [beamtalk, runtime]}),
+    Base = beamtalk_repl_protocol:base_response(Msg),
+    iolist_to_binary(json:encode(Base#{<<"sites">> => [], <<"status">> => [<<"done">>]})).
```
[the remark this answers is not in the document]
Comment (jamesc) [reply runtime/apps/beamtalk_workspace/src/beamtalk_repl_ops_nav.erl]: Fixed in 81d418a — the non-list defensive fallback now returns the empty result in the field matching the requested shape (`implementors` for an `implementorsOf` query, `sites` otherwise), so the wire contract stays consistent.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/beamtalk-core/src/language_service/runtime_nav.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+/// resolved line lies past the end of `source`.
+#[must_use]
+pub fn method_relative_line_span(
+    source: &str,
+    method_span: Span,
+    relative_line: u32,
+) -> Option<Span> {
+    let header_line =
+        Position::from_byte_offset(source, ByteOffset::new(method_span.start()))?.line;
+    let target_line = header_line + relative_line.saturating_sub(1);
```
[the remark this answers is not in the document]
Comment (jamesc) [reply crates/beamtalk-core/src/language_service/runtime_nav.rs]: Fixed in 2af9779 — `method_relative_line_span` now returns `None` for `relative_line == 0` (the contract is 1-based) instead of mapping it to the header via `saturating_sub`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/beamtalk-lsp/src/server.rs [hunk 4]
```diff
@@ -1527,6 +1647,30 @@ fn path_to_stdlib_uri(path: &Utf8PathBuf) -> Option<Url> {
...
+    sites
+        .iter()
+        .filter_map(|site| {
+            let loc = svc.locate_nav_site(&site.class, &site.selector, site.line)?;
+            let source = svc.file_source(&loc.file)?;
+            let range = span_to_range(loc.span, &source);
+            Some(tower_lsp::lsp_types::Location {
+                uri: path_to_uri(&loc.file)?,
+                range,
+            })
```
[the remark this answers is not in the document]
Comment (jamesc) [reply crates/beamtalk-lsp/src/server.rs]: Valid — deferred and tracked in BT-2264 (add an indexed `(class, selector, class_side)` lookup so `locate_nav_site` isn't a per-site full scan, plus a borrowed source accessor to avoid cloning each file). Acceptable for this opt-in foundation; BT-2240 is what turns the runtime path into a hot path.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by jamesc** [comment]
@copilot resolve comments

**Comment by jamesc** [comment]
Closing . Went different direction
